test/integration/log_system.cc

```diff
@@ -814,98 +834,84 @@ TEST_F(LogSystemTest, RecordAndPlayback)
-          << _playedMsg.pose(i).name();
-      EXPECT_NEAR(posePlayed.Rot().Yaw(), poseRecorded.Rot().Yaw(), 0.1)
-          << _playedMsg.pose(i).name();
-    }
+        // Loop through all recorded poses, and check them against the
+        // playedback poses.
+        for (int i = 0; i < recordedMsg.pose_size(); ++i)
+        {
+          const math::Pose3d &poseRecorded = msgs::Convert(recordedMsg.pose(i));
+          const std::string &name = recordedMsg.pose(i).name();
+
+          auto entity = _ecm.EntityByComponents(
+              components::Name(recordedMsg.pose(i).name()));
+          ASSERT_NE(kNullEntity, entity);
+          auto poseComp = _ecm.Component<components::Pose>(entity);
+          ASSERT_NE(nullptr, poseComp);
+          const auto &posePlayed = poseComp->Data();
+
+          EXPECT_NEAR(posePlayed.Pos().X(), poseRecorded.Pos().X(), 0.1)
+            << name;
+          EXPECT_NEAR(posePlayed.Pos().Y(), poseRecorded.Pos().Y(), 0.1)
+            << name;
+          EXPECT_NEAR(posePlayed.Pos().Z(), poseRecorded.Pos().Z(), 0.1)
+            << name;
+
+          EXPECT_NEAR(posePlayed.Rot().Roll(), poseRecorded.Rot().Roll(), 0.1)
+            << name;
+          EXPECT_NEAR(posePlayed.Rot().Pitch(), poseRecorded.Rot().Pitch(), 0.1)
+            << name;
+          EXPECT_NEAR(posePlayed.Rot().Yaw(), poseRecorded.Rot().Yaw(), 0.1)
+            << name;
+        }
 
-    ++recordedIter;
-    nTotal++;
-  };
+        ++recordedIter;
+      });
 
-  // Subscribe to ignition topic and compare to logged poses
-  transport::Node node;
-  // TODO(louise) The world name should match the recorded world
-  node.Subscribe("/world/default/dynamic_pose/info", msgCb);
-
-  int playbackSteps = 500;
-  int poseHz = 60;
-  int expectedPoseCount = playbackSteps * 1e-3 / (1.0/poseHz);
-  // Run for a few seconds to play back different poses
-  playServer.Run(true, playbackSteps, false);
-
-  int sleep = 0;
-  int maxSleep = 30;
-  for (; nTotal < expectedPoseCount && sleep < maxSleep; ++sleep)
-  {
-    std::this_thread::sleep_for(std::chrono::milliseconds(100));
-  }
-  EXPECT_NE(maxSleep, sleep);
+  playServer.AddSystem(playbackPoseTester.systemPtr);
+
+  // Playback a subset of the log file and check for poses. Note: the poses are
+  // checked in the playbackPoseTester
+  playServer.Run(true, 500, false);
+
+  // Count the total number of pose messages in the log file
+  for (auto it = batch.begin(); it != batch.end(); ++it, ++nTotal) { }
 
-  #if !defined (__APPLE__)
-  /// \todo(anyone) there seems to be a race condition that sometimes cause an
-  /// additional messages to be published by the scene broadcaster
-  // 60Hz
-  EXPECT_TRUE(nTotal == expectedPoseCount || nTotal == expectedPoseCount + 1)
+  // The expectedPoseCount might be off by ±2 because the ign transport's
+  // throttle mechanism (which is used by the SceneBroadcaster when publishing
+  // poses) uses real-time
+  EXPECT_LE(std::abs(nTotal - expectedPoseCount), 2)
```

Comment:
I guess 2 may still be conservative depending on the RTF, right? Not sure if there's much more we can do here though. Did you consider subscribing to the pose topic while recording?

Reply:
My hope is that it won't be affected by the RTF. I've tested it with `cpulimit` at 10% several times and it passes locally. Also, the ±2 should actually be ±1 since I'm checking if it's less than 2.

I hadn't considered subscribing to the pose topic while recording because I was trying to find a way to calculate the pose count without relying on ign-transport. This way the expected and actual counts are from two different sources and would be more likely to catch errors if the log recorder, which depends on ign-transport, failed to record the expected number of poses.

Reply:
Oops, never mind on the ±1, we're actually checking if it's `<= 2`

Reply:
>  counts are from two different sources and would be more likely to catch errors if the log recorder, which depends on ign-transport, failed to record the expected number of poses.

Fair enough. I think the current solution is good :+1: